### ISSUE TITLE
Unexpected behavior when using the same metric object in training and validation

### ISSUE THREAD

**Issue by deng-cy** [Lightning-AI/metrics#1450]
## 🐛 Bug

I find that, if using the same metric object in both training and validation model, the result is very weird. 

#### Code sample
Using the BoringModel:
```py
from pytorch_lightning.demos.boring_classes import BoringDataModule
import torch
from torch import Tensor
from pytorch_lightning import LightningModule
import pytorch_lightning as pl
import torchmetrics
class BoringModel(LightningModule):
    def __init__(self) -> None:
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.accuracy=torchmetrics.classification.Accuracy('binary')

    def forward(self, x: Tensor) -> Tensor:
        return self.layer(x)

    def loss(self, batch: Tensor, preds: Tensor) -> Tensor:
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(preds, torch.ones_like(preds))

    def training_step(self, batch: Tensor, batch_idx: int):
        output = self(batch)
        loss = self.loss(batch, output)
        self.accuracy(torch.ones_like(output),torch.ones_like(output))
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()
        self.log('train', self.accuracy)

    def validation_step(self, batch: Tensor, batch_idx: int):
        output = self(batch)
        loss = self.loss(batch, output)
        self.accuracy(torch.zeros_like(output),torch.ones_like(output))
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x["x"] for x in outputs]).mean()
        self.log('val', self.accuracy)


    def test_step(self, batch: Tensor, batch_idx: int):
        output = self(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]

model = BoringModel()
dm = BoringDataModule()

trainer = pl.Trainer()
trainer.fit(model, datamodule=dm)

```

### Expected behavior

Ideally, 'train' should give 1 and 'val' should give 0. Yet I observed 'train'=0, 'val'=0.5. There is some interference between train and val. 

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): 0.11.0
- Python & PyTorch Version (e.g., 1.0): 1.12.0
- Any other relevant information such as OS (e.g., Linux): Windows 11

### Additional context

I saw the documentation typically creates two objects even for the same metric. Is there a reason?

**Comment by justusschock** [comment]
Hey, this is actually expected and even stated in [our docs](https://torchmetrics.readthedocs.io/en/stable/pages/lightning.html#common-pitfalls).

Modular metrics do have internal states so you should make very sure you don't mix these between training and validation which means you need a separate metric for training and validation.

**Comment by deng-cy** [comment]
@justusschock  Hi, thanks for your reply. I did not see "The same holds for using seperate metrics for training, validation and testing." when I read it the first time. Can you slightly elaborate what internal state is stored for each dataloader to help me understand the mechanism? Intuitively I feel it should work as long as it's reset.
